### PR TITLE
Integrate final 2023 EIA 860 data

### LINKED ISSUE #3684: Integrate EIA860 2023 Final Release Data
### New year of data integration check-list:

Based on the [Existing Data Updates Docs](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/existing_data_updates.html)

```[tasklist]
## Data Update Checklist
- [x] [Obtain fresh data](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/existing_data_updates.html#obtain-fresh-data)
- [x] [Map the structure of the new data](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/existing_data_updates.html#map-the-structure-of-the-new-data)
- [x] [Test data extraction](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/existing_data_updates.html#test-data-extraction)
- [x] [Update table and column transformations](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/existing_data_updates.html#update-table-column-transformations)
- [x] [Update the PUDL db schema](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/existing_data_updates.html#update-the-pudl-db-schema)
- [x] [Connect datasets](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/existing_data_updates.html#connect-datasets)
- [x] [Run the ETL](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/existing_data_updates.html#run-the-etl)
- [x] [Update the output routines and run full tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/existing_data_updates.html#update-the-output-routines-and-run-full-tests)
- [x] [Run and update data validations](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/existing_data_updates.html#run-and-update-data-validations)
- [x] [Update the documentation](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/existing_data_updates.html#update-the-documentation)
- [x] SPECIAL: Go check to see if you can remove `so2_equipment_type_4` from the `fast_run_null_cols` in the `fgd_equipment_null_check` asset_check - no, still just 2010 and 2011 data
- [x] Remove two junky generator IDs from storage tabs in from `drop_all_null_records_with_multiindex()` method in `_out_eia__yearly_generators`
- [x] Update docs to capture utility ID EIA matching method
- [x] Check Christina's notebook into devtools
- [x] SPECIAL: Check in to see if the missing Early release generators came back (see notes below)
- [x] SPECIAL: Investigate record loss in `gf_eia923` - ValueError: gf_eia923: found 3074782 rows, expected 3074806.
```

## Check for missing early release generators
When integrating the 2023 860 early release data (#3676), we were effectively swapping the ER data for the 860m data which resulted in generator records were no longer being reported in the raw data. Here is a pickled df of the generators that were missing from the 2023 based ER `_out_eia__yearly_generators` table that did show up in the 2023 860m table:

[missing_from_2023_er.pkl.zip](https://github.com/user-attachments/files/15914469/missing_from_2023_er.pkl.zip)

```
from dagster import AssetKey
from pudl.etl import defs
import pandas as pd

asset_key = "_out_eia__yearly_generators"
df = defs.load_asset_value(AssetKey(asset_key))
missing_gens = pd.read_pickle('missing_from_2023_er.pkl')
sum(df.set_index(['report_date', 'plant_id_eia', 'generator_id']).index.isin(missing_gens.index)) # 1
```
Only one of these lost plants is found.

Do these records show up under other generator IDs? Only three additional records appear with an exact match on `report_date`, `plant_id_eia`, and `capacity_mw`, and they all belong to the same plant which appears to have multiple generators with identical capacity. Not an obvious match here.

So what could be going on?

**Changes in generator ID:** see, for e.g., plant ID 1439, which switches from reporting for generators 9A, 9B etc., to generators 9, 10, 11. See below, as this is likely to be a small part of the problem.

**Missing retired plants:** 922/931 of the `plant_id_eia`, `report_date` EIA combinations reported in `missing_gens.pkl` aren't found in the final EIA data. Of the 922 plant IDs in `missing_gens.pkl`, 917 plants are reported as retired and don't show up in the final EIA 860 data.

The [EIA 860 documentation](https://www.eia.gov/electricity/data/eia860/) notes that:

> The retired tab only includes those retired generators which were reported in the most current data cycle. This is not a comprehensive list. Starting with March 2017 data, [Preliminary Monthly Electric Generator Inventory](https://www.eia.gov/electricity/data/eia860m/) includes a comprehensive list of generators which retired since 2002. The list can be found on the Retired tab of the data file.

So, these generators _should_ show up under the latest EIA 860M data. 

```
df = df.set_index(['generator_id', 'plant_id_eia'])
df = df.loc[df.report_date>'2023-01-01'] # Only keep the latest data
missing_gens = missing_gens.reset_index().set_index(['generator_id', 'plant_id_eia'])
missing_gens[~missing_gens.index.isin(df.index)].index.unique()
```

Only 33 generators don't. Of these, 23 are explained by name changes to the generator in `plant_id_eia` 1439 (e.g., 9A and 9B are reported as 9 in the final data), and `plant_id_eia` 66933 has a generator named "TEST" that is clearly a placeholder. This leaves 9 generators "missing" from the data in some format, either because of other changes to generator ID reported, or because of other data input errors. This seems comfortably within the margin of error for this dataset.

## Investigate record loss in `gf_eia923`

```
from dagster import AssetKey
from pudl.etl import defs
import pandas as pd

asset_key = "out_eia923__monthly_generation_fuel_combined"
df = defs.load_asset_value(AssetKey(asset_key))
old_df = pd.read_parquet("s3://pudl.catalyst.coop/nightly/out_eia923__monthly_generation_fuel_combined.parquet")

old_df = old_df.set_index(['report_date', 'plant_id_eia'])
df = df.set_index(['report_date', 'plant_id_eia'])

old_df[~old_df.index.isin(df.index)]
```

Of the 24 records lost, all are the monthly 2023 records for `plant_id_eia` 67355 (Exeter Bulk Plant Array) for two different prime movers (`CP` and `PV`).

### PR DESCRIPTION
**Pull request by e-belfer**
# Overview

Closes #3684.

## What problem does this address?
Integrates EIA 860 2023 final data.

## What did you change?
Steps are tracked in #3684. Notable changes:

- drop two non-existent generators we were removing in the EIA early release data, which we ingested from the direct support columns in #3695
- update the documentation for the PUDL ID mapping and existing data updates process
- add a notebook @cmgosnell made to help with the ID mapping process to `devtools`

# Documentation

Make sure to update relevant aspects of the documentation.

```[tasklist]
- [x] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/nightly/release_notes.html): reference the PR and related issues.
- [x] Review and update any other aspects of the documentation that might be affected by this PR.
```

# Testing

## How did you make sure this worked? How can a reviewer verify this?

```[tasklist]
# To-do list
- [x] Run `make pytest-coverage` locally to ensure that the merge queue will accept your PR.
- [x] Review the PR yourself and call out any questions or issues you have.
- [x] Alternatively, run the `build-deploy-pudl` GitHub Action manually.
```

### CODE REVIEW

File: docs/release_notes.rst [hunk 1]
```diff
@@ -9,6 +9,10 @@ v2024.X.x (2024-XX-XX)
 New Data Coverage
 ^^^^^^^^^^^^^^^^^
 
+EIA 923
+~~~~~~~
+* Added EIA 923 final release data from 2023. See :issue:`3684` and PR :pr:`3871`.
+
```
Comment (zaneselvans) [inline docs/release_notes.rst:15]: This is the EIA-860 right? Maybe you had 923 on the brain this morning with the archiver false alarm.

File: src/pudl/output/eia.py [hunk 2]
```diff
@@ -145,18 +145,16 @@ def _out_eia__yearly_generators(
     )
 
     # If any generator data is completely empty, drop it.
-    # These are five known generators that originate from harvesting the plant and
+    # These are three known generators that originate from harvesting the plant and
```
Comment (cmgosnell) [inline src/pudl/output/eia.py:148]: 5->3 🎉

File: src/pudl/package_data/glue/utility_id_pudl.csv
Comment (cmgosnell) [inline src/pudl/package_data/glue/utility_id_pudl.csv:1]: do you know why this file diff is so big? it looks like maybe the capacity column got added in there as an unnamed column. Maybe that's it? if so it'd be good to delete it.

**Review by cmgosnell — CHANGES_REQUESTED** [review]
Looks chill! the docs updates are a good change here. the row count updates look good. my only request is about utility_id_pudl.csv. plus to update the form in the release notes!

File: docs/release_notes.rst (hunk 1 above)
Comment (e-belfer) [reply docs/release_notes.rst:15]: Yes, 100% a foolish copy-paste error I'll fix!

File: src/pudl/package_data/glue/utility_id_pudl.csv
Comment (e-belfer) [reply src/pudl/package_data/glue/utility_id_pudl.csv:1]: I found a few matches for older PUDL IDs during the matching process, so there's a large reassignment of the PUDL ID # as a result. Also noted that all these records were added as "EIA 860 Final Release 2023" in the notes column

Comment (cmgosnell) [reply src/pudl/package_data/glue/utility_id_pudl.csv:1]: Hm but before this pr this CSV does not have a notes column. and it looks like there are just a handful of records in this new unnamed column with a float that looks like the capacity - which we normally don't keep in here.

Comment (cmgosnell) [reply src/pudl/package_data/glue/utility_id_pudl.csv:1]: I just quickly downloaded the csv and it looks like its just 75 rows in that last column.

Comment (e-belfer) [reply src/pudl/package_data/glue/utility_id_pudl.csv:1]: Ah sorry, I was looking at the xlsx and not the CSV! You're totally right, I'll remove it.

File: src/pudl/package_data/glue/utility_id_pudl.csv [hunk 3]
```diff
@@ -4814,7 +4814,7 @@ utility_id_pudl,utility_id_ferc1,utility_name_ferc1,utility_id_eia,utility_name_
 4824,,,60245,Three Peaks Power LLC
 4825,,,60258,Paulding Wind Farm III LLC
 4826,,,60330,"SR South Loving, LLC"
-4827,,,60352,"Golden Fields Solar I, LLC"
+4316,,,60352,"Golden Fields Solar I, LLC"
```
Comment (cmgosnell) [inline src/pudl/package_data/glue/utility_id_pudl.csv:4817]: this is a bit of a nit but it looks like this one is also up above in the file. can we just remove it?

**Review by cmgosnell — APPROVED** [review]
one nit but it looks great ty for csv cleaning!!

**Pull request merged**